metrics/isolator.cpp

```diff
@@ -0,0 +1,379 @@
+    }
+
+    // Set `serviceEndpoint` based on flags.
+    ASSERT(flags.service_endpoint.isSome());
+    serviceEndpoint = flags.service_endpoint.get();
+  }
+
+  // Let the metrics service know about the container being launched.
+  // In the response, grab the STATSD_UDP_HOST and STATSD_UDP_PORT
+  // pair being returned and set it in the environment of the
+  // `ContainerLaunchInfo` returned from this function.
+  virtual Future<Option<ContainerLaunchInfo>> prepare(
+      const ContainerID& containerId,
+      const slave::ContainerConfig& containerConfig)
+  {
+    // For debug containers, we do not make API requests or inject new
+    // environment variables. They will inherit the environment of their
+    // parent container.
+    if (containerConfig.container_class() == ContainerClass::DEBUG) {
+      return ContainerLaunchInfo();
+    }
+
+    ContainerStartRequest containerStartRequest;
+    containerStartRequest.set_container_id(containerId.value());
+
+    return send(containerStartRequest)
+      .onAny(defer(
+        self(),
+        [=](const Future<http::Response>& response) -> Future<http::Response> {
+          if (!response.isReady()) {
+            return Failure("Error posting 'containerStartRequest' for"
+                           " container '" + containerId.value() + "': " +
+                           (response.isFailed() ?
+                               response.failure() : "Future discarded"));
+          }
+
+          return response.get();
+        }))
+      .then(defer(
+        self(),
+        [=](const http::Response& response)
+          -> Future<Option<ContainerLaunchInfo>> {
+          if (response.code == http::Status::NO_CONTENT) {
+            return None();
+          }
+
+          if (response.code != http::Status::CREATED) {
+            return Failure("Received unexpected response code "
+                           " '" + stringify(response.code) + "' when"
+                           " posting 'containerStartRequest' for container"
+                           " '" + containerId.value() + "'");
+          }
+
+          Try<ContainerStartResponse> containerStartResponse =
+            parse<ContainerStartResponse>(response.body);
+          if (containerStartResponse.isError()) {
+            return Failure("Error parsing the 'ContainerStartResponse' body for"
+                           " container '" + containerId.value() + "': " +
+                           containerStartResponse.error());
+          }
+
+          // TODO(klueska): For now, we require both `statsd_host` and
+          // `statsd_port` to be set. This may not be true in the future.
+          // We need to revisit this if/when this changes.
+          if (!containerStartResponse->has_statsd_host() ||
+              !containerStartResponse->has_statsd_port()) {
+            const string message = containerStartResponse.isError()
+              ? containerStartResponse.error()
+              : string();
```

Comment:
I think this ternary is unnecessary since you check for `containerStartResponse.isError()` in the previous conditional and return a Failure.